sql/src/main/java/io/crate/planner/operators/FetchOrEval.java

```diff
@@ -149,68 +225,121 @@ private Plan planWithFetch(Planner.Context plannerContext, Plan plan, List<Symbo
+    private DocTableRelation resolveDocTableRelation(Symbol output) {
+        Symbol mapped = output;
+        Symbol old;
+        do {
+            old = mapped;
+            if (old instanceof Field && ((Field) old).relation() instanceof QueriedDocTable) {
+                return ((QueriedDocTable) ((Field) old).relation()).tableRelation();
+            }
+        } while ((mapped = source.expressionMapping().get(old)) != null);
+        throw new IllegalStateException("Couldn't retrieve DocTableRelation from " + output);
+    }
 
-                return i;
+    private static Symbol toInputColOrFetchRef(Symbol output,
+                                               List<Symbol> sourceOutputs,
+                                               Map<DocTableRelation, InputColumn> fetchInputColumnsByTable,
+                                               BiConsumer<DocTableRelation, Reference> allocateFetchRef,
+                                               Map<Symbol, Symbol> expressionMapping) {
+        int idxInSource = sourceOutputs.indexOf(output);
+        if (idxInSource > -1) {
+            return new InputColumn(idxInSource, sourceOutputs.get(idxInSource).valueType());
+        }
+        return FieldReplacer.replaceFields(output, f -> {
+            int idx = sourceOutputs.indexOf(f);
+            if (idx > -1) {
+                return new InputColumn(idx, sourceOutputs.get(idx).valueType());
             }
+            AnalyzedRelation relation = f.relation();
+            if (relation instanceof QueriedDocTable) {
+                DocTableRelation docTableRelation = ((QueriedDocTable) relation).tableRelation();
+
+                Symbol symbol = expressionMapping.get(f);
+                return RefReplacer.replaceRefs(symbol, ref -> {
+                    if (ref.granularity() == RowGranularity.DOC) {
+                        ref = DocReferences.toSourceLookup(ref);
+                    }
+                    allocateFetchRef.accept(docTableRelation, ref);
+                    InputColumn fetchId = fetchInputColumnsByTable.get(docTableRelation);
+                    assert fetchId != null : "fetchId InputColumn for " + docTableRelation + " must be present";
+                    return new FetchReference(fetchId, ref);
+                });
+            }
+            Symbol mapped = expressionMapping.get(output);
+            assert mapped != null
+                : "Field mapping must exists for " + output + " in " + expressionMapping;
+            return toInputColOrFetchRef(
+                mapped, sourceOutputs, fetchInputColumnsByTable, allocateFetchRef, expressionMapping);
+        });
+    }
+
+    private Plan planWithEvalProjection(Planner.Context plannerContext, Plan plan, List<Symbol> sourceOutputs) {
+        // TODO: isLastFetch check here is probably not safe - would need to make sure
+        // symbols used in orderBy are not removed, or worse shuffled
```

Comment:
How can symbols be removed or shuffled? If it is not safe, could we omit the check and just return the plan here?

Reply:
Ah I think the comment is a bit missleading. The case where `isLastFetch == true` **is** safe. It's the other case that might not be safe.

I'll address this.